File: Java/bst/src/bst/Bst.java
/* 
 * Rafael Alcalde Azpiazu: rafael.alcalde.azpiazu (rafael.alcalde.azpiazu@udc.es)
 * Eva Suárez García: eva.suarez.garcia (eva.suarez.garcia@udc.es)
 */

package bst;

public class Bst {
	// TODO deberíamos plantearnos poner un método arbolVacio de todos modos?
	// Rollo poner a nulo el nodo y via (aunque parece un desperdicio de memoria tremendo)
	// Más bien poner todos los atributos del nodo a null
	private Node root = null;
	
	public Integer root() {
		return this.root.getKey();
	}
	
	public Bst rightChild() {
		Bst rightChild = new Bst();
		rightChild.root = this.root.getRightChild();
		return rightChild;
	}
	
	public Bst leftChild() {
		Bst leftChild = new Bst();
		leftChild.root = this.root.getLeftChild();
		return leftChild;
	}

	public boolean isEmptyTree() {
		return this.root == null;
	}

	// TODO usar nodo padre para poder tener node == null
	// TODO Pasar parentNode como árbol permite insertar la raíz al tener un puntero indirecto a ella
	// TODO inicialmente, node = root, parentNode = this
	private void insertR(Node node, Bst parentNode, Integer key) {
		// Insert in empty tree
		if (node == null) {
			// Insert root of the complete tree
			if (parentNode.root == null) { 
				parentNode.root = new Node();
				parentNode.root.setKey(key);
			// Insert regular node
			} else if (parentNode.root.getLeftChild() == node) {
				parentNode.root.setLeftChild(new Node());
				parentNode.root.getLeftChild().setKey(key);
			} else {
				parentNode.root.setRightChild(new Node());
				parentNode.root.getRightChild().setKey(key);
			}
		// Insert in left tree child TODO preguntar a Fer si es correcto
		} else if (key < node.getKey()) {
			Bst aux = new Bst();
			aux.root = node;
			insertR(node.getLeftChild(), aux, key);
		// Insert in right tree child
		} else if (key > node.getKey()) {
			Bst aux = new Bst();
			aux.root = node;
			insertR(node.getRightChild(), aux, key);
		}
		// Duplicates are ignored
	}
	
	private void insertI(Integer key) {
		Node newNode = new Node();
		newNode.setKey(key);
		
		// Insert in empty tree
		if (this.root == null) { 
			this.root = newNode;
		} else {
			Node parent = null;
			Node child = this.root;
			
			// Search for the key's right place in the tree
			while ((child != null) && (child.getKey() != key)) {
				parent = child;
				if (key < child.getKey()) {
					child = child.getLeftChild();
				} else {
					child = child.getRightChild();
				}
			}
			
			// Insert new key in its place
			if (child == null) {
				if (key < parent.getKey()) {
					parent.setLeftChild(newNode);
				} else {
					parent.setRightChild(newNode);
				}
			}
			// Duplicates are ignored
		}
	}
	
	public void insertKey(Integer key) {
		insertI(key);
	}
	
	private Bst searchR(Node node, Integer key) {
		Bst result;
		
		if (node == null) {
			result = new Bst(); 
		} else if (key == node.getKey()) {
			result = new Bst();
			result.root = node;
		} else if (key < node.getKey()) {
			result = searchR(node.getLeftChild(), key);
		} else {
			result = searchR(node.getRightChild(), key);
		}
		
		return result;
	}
	// TODO la definición mal hecha del árbol vacío también hace que el resultado al no encontrar
	// una clave sea distinto entre la versión iterativa y la recursiva
	private Bst searchI(Integer key) {
		Node node = this.root;
		
		while ((node != null) && (node.getKey() != key)) {
			if (key < node.getKey()) {
				node = node.getLeftChild();
			} else {
				node = node.getRightChild();
			}
		}
		
		Bst tree = new Bst();
		tree.root = node;
		return tree;
	}
	
	public Bst searchKey(Integer key) {
		return searchI(key);
	}
	
	// del: node with the key to delete
	// parentNode: parent of the node that will replace del
	// This method replaces the node to remove with the node with
	// the greatest key from its left subtree.
	private void deleteAux(Node del, Node parentNode, Node node) {
		// We need the parent node instead of the node itself
		// in order to take the children of the node that
		// will replace del and put them as children of
		// parentNode (<- TODO para la memoria más que nada)
		if (node.getRightChild() != null) { 
			deleteAux(del, node, node.getRightChild());
		} else {
			del.setKey(node.getKey());
			parentNode.setRightChild(node.getLeftChild());
		}
	}
	// TODO si se pasa a clase estática, es posible que parentNode necesite ser un Bst como en la
	// inserción recursiva
	// parentNode: parent of the node to be deleted. Used to reorder the children of the node
	// to be deleted.
	private void deleteR(Node node, Node parentNode, Integer key) {
		if (node != null) {
			if (key < node.getKey()) {
				deleteR(node.getLeftChild(), node, key);
			} else if (key > node.getKey()) {
				deleteR(node.getRightChild(), node, key);
			} else {
				// Delete node with at most one child
				if (node.getLeftChild() == null) {
					// Delete root of the total tree
					if (parentNode == null) {
						this.root = node.getRightChild();
					// Check if node was a right or a left child
					} else if (parentNode.getLeftChild() == node) {
						parentNode.setLeftChild(node.getRightChild());
					} else {
						parentNode.setRightChild(node.getRightChild());
					}
				} else if (node.getRightChild() == null) {
					// Delete root of the total tree
					if (parentNode == null) {
						this.root = node.getLeftChild();
					// Check if node was a right or a left child
					} if (parentNode.getLeftChild() == node) {
						parentNode.setLeftChild(node.getLeftChild());
					} else {
						parentNode.setRightChild(node.getLeftChild());
					}
				// Delete node with two children
				} else {
					deleteAux(node, node, node.getLeftChild());
				}
			}
		}
	}
	
	private void deleteI(Integer key) {
		int numChildren;
		Node rm;			// Node to remove
		Node parentRm;		// Parent of the node to remove
		Node notEmptyChild;	// If the node to remove has only one child, the one that isn't empty
		Node maxLeftChild;	// If the node to remove has two children, the node with the greatest key of the left subtree
		// TODO nombre de notEmptyChild
		parentRm = null;
		rm = this.root;
		
		// Search for the node to remove
		while ((rm != null) && (rm.getKey() != key)) {
			parentRm = rm;
			
			if (key < rm.getKey()) { 
				rm = rm.getLeftChild();
			} else { 
				rm = rm.getRightChild();
			}
		}

		// If the key isn't in the tree, don't do anything
		if (rm != null) {
			numChildren = 0;
			if (rm.getLeftChild() != null) 
				numChildren++;
			if (rm.getRightChild() != null) 
				numChildren++;
			
			switch (numChildren) {
				// Remove leaf node
				case 0:
					if (parentRm == null) {
						this.root = null; // TODO no corresponde con la definición de árbol vacío del constructor
					} else if (parentRm.getLeftChild() == rm) {
						parentRm.setLeftChild(null);
					} else {
						parentRm.setRightChild(null);
					}
					break;
		
				// Remove node with one child
				case 1:
					if (rm.getLeftChild() == null) {
						notEmptyChild = rm.getRightChild();
					} else {
						notEmptyChild = rm.getLeftChild();
					}
					
					if (parentRm == null) {
						this.root = notEmptyChild;
					} else if (parentRm.getLeftChild() == rm) {
						parentRm.setLeftChild(notEmptyChild);
					} else {
						parentRm.setRightChild(notEmptyChild);
					}
					break;
			
				// Remove node with two children
				case 2:
					parentRm = rm;
					maxLeftChild = rm.getLeftChild();
					
					// Search for the node with the greatest key of the left subtree
					while (maxLeftChild.getRightChild() != null) {
						parentRm = maxLeftChild;
						maxLeftChild = maxLeftChild.getRightChild();
					}
					
					// Replace the key of the node to remove with the found one
					rm.setKey(maxLeftChild.getKey());
				
					// TODO Put maxLeftChild's children as its parent's
					if (parentRm == rm) {
						// If maxLeftChild doesn't have right child
						parentRm.setLeftChild(maxLeftChild.getLeftChild());
					} else {
						parentRm.setRightChild(maxLeftChild.getLeftChild());
					}
					break;
			}
		}
	}
	
	public void deleteKey(Integer key) {
		deleteR(this.root, null, key);
	}
}

File: OCaml/bst.ml
(*
 * Rafael Alcalde Azpiazu: rafael.alcalde.azpiazu (rafael.alcalde.azpiazu@udc.es)
 * Eva Suárez García: eva.suarez.garcia (eva.suarez.garcia@udc.es)
 *)

type bst = Empty | Node of int * bst ref * bst ref;;

let emptyTree () = ref Empty;;

let leftChild tree = match !tree with
    Node(_,left,_) -> left;;

let rightChild tree = match !tree with
    Node(_,_,right) -> right;;

let root tree = match !tree with
    Node(key,_,_) -> key;;

let isEmptyTree tree = match !tree with
 	Empty -> true
	| _ -> false;;

let rec insertR tree key =
	match !tree with
		Empty -> tree := Node(key, ref Empty, ref Empty)
		| Node(k,left,right) -> if (key < k) then
                                    insertR left key
				 				else if (key > k) then
                                    insertR right key;;
                                (* Duplicates are ignored *)

let insertI tree key =
	let newNode = ref (Node (key, (ref Empty), (ref Empty))) in
	(* Insert in empty tree *)
	if (!tree = Empty) then
		tree := !newNode
	else
		let parent = ref Empty in
		let child = ref (!tree) in
		(* Search for the key's right place in the tree *)
		while (((!child) <> Empty) && ((root child) <> key)) do
			parent := !child;
			if (key < (root child)) then
				child := !(leftChild child)
			else
				child := !(rightChild child)
		done;

		if (!child = Empty) then
			if (key < (root parent)) then
				(leftChild parent) := !newNode
			else
				(rightChild parent) := !newNode;;

let insertKey =
	insertR;;
	(*insertI;;*)

let rec searchR tree k =
	match !tree with
		Empty -> ref Empty
		| Node(key,left,right) -> if (k = key) then
                                    ref (Node(key,left,right))
								  else if (k < key) then
                                    searchR left k
								  else
                                    searchR right k;;

let searchI tree key =
    let node = ref !tree in
    while (!node <> Empty) && ((root node) <> key) do
        if key < (root node) then
            node := !(leftChild node)
        else
            node := !(rightChild node)
    done;
    node;;

let searchKey =
	searchR;;
	(*searchI;;*)

let rec deleteR tree key =
    (* This function is first called only when 'node' has 2 children, and
       recursive calls only happen if this argument is going to be non empty,
       so we don't need to check if 'node' is empty before calling rightChild *)
    let rec delAux node parentNode =
		match !(rightChild node) with
			Node(key,left,right) -> delAux (rightChild node) node
			|Empty -> tree := Node(root node, leftChild tree, rightChild tree);
					  (* Check if the left subtree didn't have a right child *)
					  (*TODO comentar a Rafa si hace falta que !parentNode = !node no funciona*)
					  if (root parentNode) = (root node) then
					  	(leftChild parentNode) := !(leftChild node)
					  else
                      	(rightChild parentNode) := !(leftChild node)
	in match !tree with
	 	Empty -> () (* The key was not in the tree *)
		|Node(k,left,right) -> if key < k then
                                    deleteR left key
                               else if key > k then
                                    deleteR right key
                               (* From this point on we have found the node with the key to delete *)
                               else if !left = Empty then
                                    tree := !right
                               else if !right = Empty then
                                    tree := !left
                               else delAux left tree;; (*TODO cambiar en la memoria, antes se llamaba con left left*)

let deleteI tree key =
	let rm = ref (!tree) in
	let parentRm = ref Empty in
	let nonEmptyChild = ref Empty in
	let maxLeftChild = ref Empty in
    let parentMaxLeftChild = ref Empty in

	(* Search for the node to remove *)
	while ((!rm <> Empty) && ((root rm) <> key)) do
		parentRm := !rm;
		if (key < (root rm)) then
			rm := !(leftChild rm)
		else
			rm := !(rightChild rm)
	done;

	(* If the key isn't in the tree, don't do anything *)
	if (!rm <> Empty) then
		let numChildren = ref 0 in
		if (!(leftChild rm) <> Empty) then
			numChildren := (!numChildren + 1);
		if (!(rightChild rm) <> Empty) then
			numChildren := (!numChildren + 1);

		match !numChildren with
		 	0 -> if (!parentRm = Empty) then
			     	tree := Empty
			     else if ((leftChild parentRm) = rm) then
				 	(leftChild parentRm) := Empty
				 else
				 	(rightChild parentRm) := Empty;
			| 1 -> if (!(leftChild rm) = Empty) then
			     	nonEmptyChild := !(rightChild rm)
				 else
				 	nonEmptyChild := !(leftChild rm);

				 if (!parentRm = Empty) then
				 	tree := !nonEmptyChild
				 else if (!(leftChild parentRm) = !rm) then
				 	(leftChild parentRm) := !nonEmptyChild
				 else
				 	(rightChild parentRm) := !nonEmptyChild;
			| 2 -> parentMaxLeftChild := !rm;
			     maxLeftChild := !(leftChild rm);
				 while (!(rightChild maxLeftChild) <> Empty) do
				 	parentMaxLeftChild := !maxLeftChild;
					maxLeftChild := !(rightChild maxLeftChild)
				 done;
                 (* The node is changed through its parent *)
                 if (!parentRm = Empty) then (* Change the root's key *)
                    tree := Node(root maxLeftChild, leftChild rm, rightChild rm)
                 (* Change a regular node's key *)
                 else if ((root rm) < (root parentRm)) then
				    (leftChild parentRm) := Node(root maxLeftChild, leftChild rm, rightChild rm)
                 else if ((root rm) > (root parentRm)) then
                    (rightChild parentRm) := Node(root maxLeftChild, leftChild rm, rightChild rm);

				 if (!parentMaxLeftChild = !rm) then
				 	(leftChild parentMaxLeftChild) := !(leftChild maxLeftChild)
				 else
				 	(rightChild parentMaxLeftChild) := !(leftChild maxLeftChild);;

let deleteKey =
	deleteR;;
	(*deleteI;;*)
